docs/types/reference-types.rst

```diff
@@ -37,6 +37,12 @@ non-persistent area where function arguments are stored, and behaves mostly like
     data location can also be returned from functions, but it is not possible to
     allocate such types.
 
+.. note::
+    Arrays and structs with ``calldata`` data location declared in a function body must be assigned before
+    use. Arrays and structs with ``calldata`` data location as the returned parameters of a function
+    must be assigned before returning. This rule is enforced by the compiler to allow complex initialization patterns.
+    A workaround is assign the variable to itself after declaration.
```

Comment:
```suggestion
    must be assigned before returning. This rule is enforced by the compiler to allow complex initialization patterns.
    A workaround is assign the variable to itself after declaration.
```
It is actually the workaround which allows complex initialization patterns. The rule enforced by the compiler is that you can't use the `calldata` arrays and structs without initialization.